modules/libcom/src/osi/os/posix/osdThread.c

```diff
@@ -682,32 +682,45 @@ void epicsThreadMustJoin(epicsThreadId id)
 {
     void *ret = NULL;
     int status;
+    int prev;
+    epicsThreadId self;
 
-    if(!id) {
+    if(!id)
         return;
-    } else if(epicsAtomicCmpAndSwapIntT(&id->joinable, 1, 0)!=1) {
-        if(epicsThreadGetIdSelf()==id) {
-            errlogPrintf("Warning: %s thread self-join of unjoinable\n", id->name);
+
+    prev = epicsAtomicCmpAndSwapIntT(&id->joinable, 1, 0);
+    self = epicsThreadGetIdSelf();
+
+    if(prev==0) {
+        /* join of unjoinable */
+        if(self==id) {
+            errlogPrintf(ERL_WARNING ": %s thread self-join of unjoinable\n", id->name);
+            return;
```

Comment:
I think you are missing a call to `free_threadInfo()`

Reply:
Or is this meant to be a complete no-op? The comment at the end of the function says "always decrement refcnt"; the "always" is confusing to me, and there are not docs.

Reply:
`prev==0` implies that `refcnt==1`.  For a self-join, this last reference is for the thread itself, and will be decremented when it actual exits.  This happens automatically through the use of a destructor function passed to `pthread_key_create()` when creating the thread local behind `epicsThreadGetIdSelf()`.